### PR TITLE
Update 3dconnexion to 10.6.1 and add appcast, uninstall, zap and caveats

### PR DESCRIPTION
**Pull request by ytbmulder**
After making all changes to the cask:

- [x] `brew cask audit --download {{cask_file}}` is error-free.
- [x] `brew cask style --fix {{cask_file}}` reports no offenses.
- [x] The commit message includes the cask’s name and version.
- [x] The submission is for [a stable version](https://github.com/Homebrew/homebrew-cask/blob/master/doc/development/adding_a_cask.md#finding-a-home-for-your-cask) or [documented exception](https://github.com/Homebrew/homebrew-cask/blob/master/doc/development/adding_a_cask.md#but-there-is-no-stable-version).

Even though this cask installs a kext, when the `uninstall kext:` key is used, `brew uninstall` works as expected (and CI passes), but `brew zap` reports the following failure:
```
$ brew cask zap --verbose ./3dconnexion.rb
==> Implied "brew cask uninstall 3dconnexion"
==> Running uninstall process for 3dconnexion; your password may be necessary
==> Removing launchctl service com.3dconnexion.nlserverIPalias
Password:
==> Unloading kernel extension com.3dconnexion.driver
(kernel) Can't unload kext com.3dconnexion.driver; classes have instances:
(kernel)     Kext com.3dconnexion.driver class ConnexionVirtualHID has 1 instance.
Failed to unload com.3dconnexion.driver - (libkern/kext) kext is in use or retained (cannot unload).
Error: Failure while executing; `/usr/bin/sudo -E -- /sbin/kextunload -b com.3dconnexion.driver` exited with 3. Here's the output:
(kernel) Can't unload kext com.3dconnexion.driver; classes have instances:
(kernel)     Kext com.3dconnexion.driver class ConnexionVirtualHID has 1 instance.
Failed to unload com.3dconnexion.driver - (libkern/kext) kext is in use or retained (cannot unload).
```
This fail is not caught in the CI flow, because it seems that `brew zap` is not tested.
For this reason, a comment inside the `uninstall kext:` key is added.

### CODE REVIEW

**Comment by vitorgalvao** [comment]
Removing the `kext` works when using `uninstall` but not when using `zap`? That is weird; the latter should invoke the former, not do anything different.

Ping @reitermarkus.

**Comment by ytbmulder** [comment]
> Removing the `kext` works when using `uninstall` but not when using `zap`? That is weird; the latter should invoke the former, not do anything different.

That is my understanding too. The following happens with the `kext:` key enabled:

```
$ cat 3dconnexion.rb | grep 'kext:'
            kext:      'com.3dconnexion.driver', # Fails to unload kext (classes have instances) when executing `brew zap`.
```

When using `uninstall`:
```
$ brew cask uninstall --verbose 3dconnexion.rb
==> Uninstalling Cask 3dconnexion
==> Running uninstall process for 3dconnexion; your password may be necessary
==> Removing launchctl service com.3dconnexion.nlserverIPalias
==> Running uninstall script /Applications/3Dconnexion/Uninstall 3Dconnexion Driver.app/Contents/Resources/rm3dcx
==> Uninstalling packages:
com.3dconnexion.3dconnexionpairing
com.3dconnexion.3dxacrobat3d.installer
com.3dconnexion.3dxmaya
com.3dconnexion.3dxnlserver
com.3dconnexion.3dxphotoshop.installer
com.3dconnexion.3dxsketchup
com.3dconnexion.3dxwaremacinfo.installer
com.3dconnexion.collage
com.3dconnexion.driver.installer
com.3dconnexion.pluginpkg.installer
com.3dconnexion.puzzle
com.3dconnexion.trainer
com.3dconnexion.viewer
==> Removing files:
/Applications/3Dconnexion
/Library/Extensions/3Dconnexion.kext
/Library/Frameworks/3DconnexionClient.framework
==> Removing directories if empty:
/Applications/3Dconnexion
==> Purging files for version 10-6-1_r2897,a8111ac7-d575-9772-1dc6-5c37a3a6c64c of Cask 3dconnexion
```

When using `zap`:
```
$ brew cask zap --verbose 3dconnexion.rb
==> Implied "brew cask uninstall 3dconnexion"
==> Running uninstall process for 3dconnexion; your password may be necessary
==> Removing launchctl service com.3dconnexion.nlserverIPalias
==> Unloading kernel extension com.3dconnexion.driver
(kernel) Can't unload kext com.3dconnexion.driver; classes have instances:
(kernel)     Kext com.3dconnexion.driver class ConnexionVirtualHID has 2 instances.
Failed to unload com.3dconnexion.driver - (libkern/kext) kext is in use or retained (cannot unload).
Error: Failure while executing; `/usr/bin/sudo -E -- /sbin/kextunload -b com.3dconnexion.driver` exited with 3. Here's the output:
(kernel) Can't unload kext com.3dconnexion.driver; classes have instances:
(kernel)     Kext com.3dconnexion.driver class ConnexionVirtualHID has 2 instances.
Failed to unload com.3dconnexion.driver - (libkern/kext) kext is in use or retained (cannot unload).
```

The following happens with the `kext:` key disabled and using `zap`:
```
$ cat 3dconnexion.rb | grep 'kext:'
            # kext:      'com.3dconnexion.driver', # Fails to unload kext (classes have instances) when executing `brew zap`.
```

```
$ brew cask zap --verbose 3dconnexion.rb
==> Implied "brew cask uninstall 3dconnexion"
==> Running uninstall process for 3dconnexion; your password may be necessary
==> Removing launchctl service com.3dconnexion.nlserverIPalias
==> Running uninstall script /Applications/3Dconnexion/Uninstall 3Dconnexion Driver.app/Contents/Resources/rm3dcx
==> Uninstalling packages:
com.3dconnexion.3dconnexionpairing
com.3dconnexion.3dxacrobat3d.installer
com.3dconnexion.3dxmaya
com.3dconnexion.3dxnlserver
com.3dconnexion.3dxphotoshop.installer
com.3dconnexion.3dxsketchup
com.3dconnexion.3dxwaremacinfo.installer
com.3dconnexion.collage
com.3dconnexion.driver.installer
com.3dconnexion.pluginpkg.installer
com.3dconnexion.puzzle
com.3dconnexion.trainer
com.3dconnexion.viewer
==> Removing files:
/Applications/3Dconnexion
/Library/Extensions/3Dconnexion.kext
/Library/Frameworks/3DconnexionClient.framework
==> Removing directories if empty:
/Applications/3Dconnexion
==> Dispatching zap stanza
==> Running zap process for 3dconnexion; your password may be necessary
==> Trashing files:
/Library/Application Support/3Dconnexion
/Library/LaunchDaemons/com.3dconnexion.nlserverIPalias.plist
/Library/PreferencePanes/3Dconnexion.prefPane
~/Library/Logs/3Dconnexion
~/Library/Preferences/3Dconnexion
~/Library/Preferences/com.3dconnexion.*.plist
~/Library/Saved Application State/com.3dconnexion.*.savedState
==> Removing all staged versions of Cask '3dconnexion'
```

File: Casks/3dconnexion.rb [hunk 1]
```diff
@@ -1,12 +1,42 @@
...
+            script:    [
+                         executable: "#{appdir}/3Dconnexion/Uninstall 3Dconnexion Driver.app/Contents/Resources/rm3dcx",
+                         sudo:       true,
+                       ],
+            delete:    [
+                         '/Applications/3Dconnexion',
+                         '/Library/Extensions/3Dconnexion.kext',
+                         '/Library/Frameworks/3DconnexionClient.framework',
+                       ],
+            rmdir:     '/Applications/3Dconnexion'
```
Comment (reitermarkus) [inline Casks/3dconnexion.rb]: This is already in `delete`.

**Comment by reitermarkus** [comment]
The `kext` issue seems to be an upstream issue. For me it is also failing with `uninstall`, not just `zap`.

And after `--force` installing it multiple times I apparently have 3 instances of the driver running:

```
$ /usr/bin/sudo -E -- /sbin/kextunload -b com.3dconnexion.driver
Password:
(kernel) Can't unload kext com.3dconnexion.driver; classes have instances:
(kernel)     Kext com.3dconnexion.driver class ConnexionVirtualHID has 3 instances.
Failed to unload com.3dconnexion.driver - (libkern/kext) kext is in use or retained (cannot unload).
```

**Comment by ytbmulder** [comment]
> The `kext` issue seems to be an upstream issue. For me it is also failing with `uninstall`, not just `zap`.

I tried a similar solution as `soundflower.rb` uses in `homebrew-cask` [here](https://github.com/Homebrew/homebrew-cask/blob/5a23fe5486d8fd757933ebb24353c270682c5b80/Casks/soundflower.rb#L19-L23), but without any luck.

Using `/usr/sbin/kextstat`, I found that this `kext` has other kexts referenced by it:
```
213    0 0xffffff7f83496000 0x10000    0x10000    com.3dconnexion.driver (1.1.0) 94E8C49C-EDED-3526-88BE-E7207560C0D2 <212 39 34 5 4 3>
```
`<212 39 34 5 4 3>` shows the index of the other kexts, of which some also have references:
```
  3  114 0xffffff7f80a04000 0x20fd0    0x20fd0    com.apple.kpi.iokit (16.7.0) AB086873-A827-4BB8-885D-53EC156842CE
  4  120 0xffffff7f80a25000 0xd200     0xd200     com.apple.kpi.libkern (16.7.0) 6C72B34A-0F27-4CFD-A1B6-A1AF5036FBAE
  5  108 0xffffff7f80a00000 0x3dd0     0x3dd0     com.apple.kpi.mach (16.7.0) 97F00961-85F9-4404-93A0-3DCC3FA6832B
 34    7 0xffffff7f812cd000 0x7a000    0x7a000    com.apple.iokit.IOHIDFamily (2.0.0) 7F020D24-38F2-3DC4-A69C-40665E78D9F4 <14 7 6 5 4 3 2 1>
 39    3 0xffffff7f81202000 0x9a000    0x9a000    com.apple.iokit.IOUSBFamily (900.4.1) 1B35F2B8-7EB1-3F3A-86DD-35F07996E335 <19 12 7 5 4 3 1>
212    1 0xffffff7f8348d000 0x9000     0x9000     com.apple.iokit.IOUSBHIDDriver (900.4.1) 00FA3E2D-73DB-3778-88DD-681AEB05EF22 <39 34 5 4 3 1>
```

I tried to manually unload a kext without any references, but that also does not work:
```
$ sudo kextunload -bundle-id com.apple.kpi.iokit
(kernel) Can't remove kext com.apple.kpi.iokit; services failed to terminate - 0xdc008018.
Failed to unload com.apple.kpi.iokit - (libkern/kext) kext is in use or retained (cannot unload).
```

**Comment by reitermarkus** [comment]
Probably the kext can only be unloaded by rebooting.

**Comment by ytbmulder** [comment]
> Probably the kext can only be unloaded by rebooting.

Sounds good to me :)

**Comment by reitermarkus** [comment]
Thanks, @ytbmulder!

**Pull request merged**